stable_baselines/common/policies.py

```diff
@@ -246,7 +318,50 @@ def value(self, obs, state=None, mask=None):
         raise NotImplementedError
 
 
-class LstmPolicy(ActorCriticPolicy):
+class StatefulActorCriticPolicy(ActorCriticPolicy):
```

Comment:
1. I was hesitant about making this a subclass of `BasePolicy` since it was lifted directly from `LstmPolicy`, which is an `ActorCriticPolicy`, and I couldn't find an example of a non actor-critic stateful policy in Stable Baselines. However, I agree it doesn't seem to depend on actor critic in any way, so am inclined to move it up. This would mean `LstmPolicy` would have multiple inheritance from `ActorCriticPolicy` and `StatefulPolicy` though.
2. See my previous comment about issues with the `stateful` property. Integrating with `BasePolicy` makes sense given we already have an `initial_state` property and `step` taking `states` and `masks`. However I feel a little odd about defining placeholders that are never used: RL algorithms will still need to special-case the training part based on if it is stateful or not, so I'm not sure this is making things easier? An alternative if we do 1. would be to delete `stateful` property entirely and just test for if policy is a subclass of `StatefulPolicy`
3. Agree this is problematic, will fix.

Will wait for your response on 1 & 2 before making changes.

Reply:
I would like to avoid multiple inheritance. I think we need to resolve [the issue above](https://github.com/hill-a/stable-baselines/pull/244#discussion_r270659219) before thinking about 1 and 2. It seems to be touching many points so we need to be careful about it.